### PR TITLE
[minor_change] Ability to add multiple templates using the resource [mso_schema]

### PR DESCRIPTION
**Pull request by shrsr**

### CODE REVIEW

File: examples/schema/main.tf [hunk 1]
```diff
@@ -14,9 +14,46 @@ provider "mso" {
   // platform = "nd"  // use when logging in ND.
 }
 
-// create data source
+// create resource with template_name and tenant_id
 resource "mso_schema" "schema1" {
   name          = "demo_schema"
   template_name = "tempu"
   tenant_id     = "0000ffff0000000000000010"
```
Comment (lhercot) [inline examples/schema/main.tf:21]: Add the tenant as a resource or a datasource to use an interpolation here.

File: mso/resource_mso_schema.go [hunk 2]
```diff
@@ -50,18 +74,35 @@ func resourceMSOSchemaCreate(d *schema.ResourceData, m interface{}) error {
 	log.Printf("[DEBUG] Schema: Beginning Creation")
 	msoClient := m.(*client.Client)
 	name := d.Get("name").(string)
-	templateName := d.Get("template_name").(string)
-	tenandId := d.Get("tenant_id").(string)
-
-	schemaApp := models.NewSchema("", name, templateName, tenandId)
+	var schemaApp *models.Schema
```
Comment (lhercot) [inline mso/resource_mso_schema.go:77]: Where do you check if neither template_name nor templates is provided?

File: mso/resource_mso_schema.go [hunk 3]
```diff
@@ -107,70 +146,111 @@ func resourceMSOSchemaUpdate(d *schema.ResourceData, m interface{}) error {
...
+				for _, valueMapOld := range getDifferenceOld {
+					valueOld := valueMapOld.(map[string]interface{})
+					for _, valueMapNew := range getDifferenceNew {
+						valueNew := valueMapNew.(map[string]interface{})
+
+						//Tenant ID has been changed. Delete Template.
+						if valueOld["name"] == valueNew["name"] && valueOld["display_name"] == valueNew["display_name"] {
+							deleteTemplate(valueOld["name"].(string), valueOld["display_name"].(string), valueOld["tenant_id"].(string), d, m)
+						}
+					}
```
Comment (lhercot) [inline mso/resource_mso_schema.go:229]: How do you take care of the templates that are removed from the main.tf are they in differenceOld? But they will not have a valueNew...

File: website/docs/r/schema.html.markdown [hunk 4]
```diff
@@ -12,6 +12,8 @@ Manages MSO Schema
 
 ## Example Usage ##
 
+### When template_name and tenant_id are used ##
```
Comment (lhercot) [inline website/docs/r/schema.html.markdown:15]: Move it after the other example and add (DEPRECATED) in the title.

File: website/docs/r/schema.html.markdown [hunk 5]
```diff
@@ -21,11 +23,39 @@ resource "mso_schema" "foo_schema" {
...
+}  
+
+```
+
 ## Argument Reference ##
 
 * `name` - (Required) name of the schema.
-* `template_name` - (Required) name of template attached to this schema.
-* `tenant_id` - (Required) tenant_id for this schema.
+* `template_name` - (Optional) **Deprecated**. name of template attached to this schema.
```
Comment (lhercot) [inline website/docs/r/schema.html.markdown:53]: name -> Name

File: website/docs/r/schema.html.markdown [hunk 6]
```diff
@@ -21,11 +23,39 @@ resource "mso_schema" "foo_schema" {
...
+* `tenant_id` - (Optional) **Deprecated**. tenant_id for this schema.
+* `template` - (Optional) template associated with this schema. Multiple templates can be created using this attribute. Type - Set.
```
Comment (lhercot) [inline website/docs/r/schema.html.markdown:55]: template -> A block that represent the template associated...
Type - Block

File: website/docs/r/schema.html.markdown [hunk 7]
```diff
@@ -21,11 +23,39 @@ resource "mso_schema" "foo_schema" {
...
+  * `name` - name of template.
```
Comment (lhercot) [inline website/docs/r/schema.html.markdown:56]: name of -> Name of

File: website/docs/r/schema.html.markdown [hunk 8]
```diff
@@ -21,11 +23,39 @@ resource "mso_schema" "foo_schema" {
...
+  * `display_name` - display name for the template
```
Comment (lhercot) [inline website/docs/r/schema.html.markdown:57]: display name -> Display name

**Review by lhercot — CHANGES_REQUESTED** [review]

File: mso/resource_mso_schema.go (hunk 2 above)
Comment (shrsr) [reply mso/resource_mso_schema.go:77]: I'll include it on my next commit

File: mso/resource_mso_schema.go (hunk 3 above)
Comment (shrsr) [reply mso/resource_mso_schema.go:229]: MSO's PATCH API works well for the templates that are removed from the configuration file. It also works well when name of the template and display name of the template are changed. It only complains when we try to change tenant id of a particular template. Hence,  when the user wants to change the tenant id of a particular template, the above code deletes that particular template and re create it.

For NDO, PATCH does not require deleting a template explicitly when the tenant id is changed. Hence we bypass the deletion of template step for NDO.

File: website/docs/r/schema.html.markdown [hunk 9]
```diff
@@ -24,8 +50,12 @@ resource "mso_schema" "foo_schema" {
 ## Argument Reference ##
 
 * `name` - (Required) name of the schema.
-* `template_name` - (Required) name of template attached to this schema.
-* `tenant_id` - (Required) tenant_id for this schema.
+* `template_name` - (Optional) **Deprecated**. Name of template attached to this schema.
```
Comment (anvitha-jain) [inline website/docs/r/schema.html.markdown:53]: attached to this schema -> attached to the schema.

File: website/docs/r/schema.html.markdown [hunk 10]
```diff
@@ -24,8 +50,12 @@ resource "mso_schema" "foo_schema" {
...
+* `tenant_id` - (Optional) **Deprecated**. tenant_id for this schema.
```
Comment (anvitha-jain) [inline website/docs/r/schema.html.markdown:54]: for this schema -> for the schema

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

File: mso/resource_mso_schema.go [hunk 11]
```diff
@@ -50,18 +76,46 @@ func resourceMSOSchemaCreate(d *schema.ResourceData, m interface{}) error {
...
+			for _, val := range template_list {
+				map_templates := make(map[string]interface{})
+				inner_templates := val.(map[string]interface{})
+				map_templates["name"] = inner_templates["name"]
+				map_templates["displayName"] = inner_templates["display_name"]
+				map_templates["tenantId"] = inner_templates["tenant_id"]
+				templates = append(templates, map_templates)
+			}
+		}
+		schemaApp = models.NewSchema("", name, "", "", templates)
```
Comment (RutvikS-crest) [inline mso/resource_mso_schema.go:109]: Hi, @shrsr NewSchema function should expect 5 arguments and should return 2 values. But here in this case, we are only catching one of the return values.

Comment (shrsr) [reply mso/resource_mso_schema.go:109]: Same as above.

File: mso/resource_mso_schema.go [hunk 12]
```diff
@@ -80,25 +134,23 @@ func resourceMSOSchemaImport(d *schema.ResourceData, m interface{}) ([]*schema.R
...
+		templates = append(templates, map_template)
 
 	}
-	if !found {
-		d.Set("template_name", "")
-		d.Set("tenant_id", "")
-	}
+	d.Set("template", templates)
+	d.Set("template_name", "")
+	d.Set("tenant_id", "")
```
Comment (lhercot) [inline mso/resource_mso_schema.go:152]: Should we set those if it is a single template?

File: mso/resource_mso_schema.go [hunk 13]
```diff
@@ -107,70 +159,198 @@ func resourceMSOSchemaUpdate(d *schema.ResourceData, m interface{}) error {
...
 
-	tempDisplayNamePayload := fmt.Sprintf(`
+		tempDisplayNamePayload := fmt.Sprintf(`
 		{
 			"op": "replace",
 			"path": "/templates/%s/displayName",
 			"value": "%s"
 		}
-	`, newTemplate, newTemplate)
+		`, newTemplate, newTemplate)
```
Comment (lhercot) [inline mso/resource_mso_schema.go:197]: Why create those even when there is no changes to them?

**Review by lhercot — CHANGES_REQUESTED** [review]
Also update to mso-go-client v1.6.0 and add vendor files

File: mso/resource_mso_schema.go (hunk 12 above)
Comment (shrsr) [reply mso/resource_mso_schema.go:152]: Yes. When we apply a configuration using multiple templates, TF state always has "template_name", "" and "tenant_id", "" set as empty.
So it is good to have those on import as well. There won't be any issues when the imported configuration is reapplied.

**Review by lhercot — DISMISSED** [review]
LGTM

File: website/docs/d/schema.html.markdown [hunk 14]
```diff
@@ -24,5 +24,5 @@ data "mso_schema" "sample_schema" {
 
 ## Attribute Reference ##
 
-* `template_name` - (Optional) name of template Attached to this schema.
-* `tenant_id` - (Optional) tenant_id for this schema.
+* `template_name` - (Optional) name of template Attached to the schema.
```
Comment (anvitha-jain) [inline website/docs/d/schema.html.markdown:27]: Attached -> attached

**Review by anvitha-jain — CHANGES_REQUESTED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Review by anvitha-jain — APPROVED** [review]
LGTM

**Pull request merged**